### PR TITLE
Resolve #304: Introduce state machine approach

### LINKED ISSUE #304: Introduce state machine approach

### PR DESCRIPTION
**Pull request by saalistaja**
- Refactored previous approach to spring state machine
- Mapped current phase, status to new onboarding state enums
- Transitions triggered by simple events from controllers and one technical event for auto-transition on status check

### CODE REVIEW

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/IdentityVerificationController.java [hunk 1]
```diff
@@ -370,41 +380,27 @@ public ObjectResponse<DocumentVerificationSdkInitResponse> initVerificationSdk(
...
         final String processId = request.getRequestObject().getProcessId();
 
-        onboardingService.verifyProcessId(ownerId, processId);
-
-        final SessionInfo sessionInfo = presenceCheckService.init(ownerId, processId);
-
-        final PresenceCheckInitResponse response = new PresenceCheckInitResponse();
-        response.setSessionAttributes(sessionInfo.getSessionAttributes());
-        return new ObjectResponse<>(response);
+        var stateMachine = stateMachineService.processStateMachineEvent(ownerId, processId, OnboardingEvent.PRESENCE_CHECK_INIT);
```
Comment (romanstrobl) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/IdentityVerificationController.java:402]: Usage of `var` is a bit inconsistent with the rest of the code. Let's keep the old approach until we switch to `var` globally.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/IdentityVerificationController.java [hunk 2]
```diff
@@ -415,24 +411,21 @@ public ObjectResponse<PresenceCheckInitResponse> initPresenceCheck(@EncryptedReq
...
         checkRequestObject(request, "resending OTP during identity verification");
 
         final OwnerId ownerId = extractOwnerId(eciesContext);
         final String processId = request.getRequestObject().getProcessId();
-        onboardingService.verifyProcessId(ownerId, processId);
 
-        IdentityVerificationEntity identityVerification = findIdentityVerification(ownerId);
-        identityVerificationOtpService.resendOtp(ownerId, identityVerification);
-        return new Response();
+        var stateMachine = stateMachineService.processStateMachineEvent(ownerId, processId, OnboardingEvent.OTP_VERIFICATION_RESEND);
```
Comment (romanstrobl) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/controller/api/IdentityVerificationController.java:427]: dtto

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/PresenceCheckService.java [hunk 3]
```diff
@@ -243,14 +260,43 @@ public Image selectPhotoForPresenceCheck(OwnerId ownerId, List<DocumentVerificat
...
+                logger.warn("Presence check failed, {}, errorDetail: '{}'", ownerId, result.getErrorDetail());
+                break;
+            case IN_PROGRESS:
+                logger.debug("Presence check still in progress, {}", ownerId);
+                break;
+            case REJECTED:
+                idVerification.setRejectReason(result.getRejectReason());
+                idVerification.setStatus(IdentityVerificationStatus.REJECTED);
+                idVerification.setTimestampLastUpdated(ownerId.getTimestamp());
+                logger.warn("Presence check rejected, {}, rejectReason: '{}'", ownerId, result.getRejectReason());
```
Comment (romanstrobl) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/PresenceCheckService.java:285]: I am not sure whether we should use a `warn` level given how frequent this error is. I would prefer using `info` level.

**Review by romanstrobl — APPROVED** [review]
In general, these changes are exactly what we needed to improve the state handling in the application. I found several minor issues which are related to consistency, several minor performance improvements and ideas. Feel free to ignore if you disagree with proposed changes.

**Comment by saalistaja** [comment]
@romanstrobl Thanks for the feedback. I have fixed all the found issues. 
Let's merge and continue later with an advanced state machine evolution.

**Pull request merged**